addons/mail/wizard/mail_compose_message.py

```diff
@@ -1032,15 +1032,19 @@ def _prepare_mail_values_dynamic(self, res_ids):
                 template_values[res_id].pop('attachment_ids', None)
                 mail_values_all[res_id].update(template_values[res_id])
 
-        # Handle recipients. Without template, if no partner_ids is given, update
-        # recipients using default recipients to be sure to notify someone
-        if not self.template_id and not self.partner_ids and email_mode:
+        # Handle recipients.
+        # - When no mail template is used and no partner_ids is given,
+        #   we fall back to the model's default recipients to guarantee delivery.
+        # - When a mail template is used but no partner_ids in rendered values,
+        #   we also fallback to default recipients to avoid missing recipients.
+        if email_mode:
             default_recipients = RecordsModel.browse(res_ids)._message_get_default_recipients()
             for res_id in res_ids:
-                mail_values_all[res_id].update(
-                    default_recipients.get(res_id, {})
-                )
-        # TDE FIXME: seems to be missing an "else" here to add partner_ids in rendering mode
+                if (
+                    (not self.template_id and not self.partner_ids)
+                    or (self.template_id and not mail_values_all[res_id].get('partner_ids'))
```

Comment:
You could be sending the template to someone you didn't want to target, just because they're the default recipient. When you specifically said you only wanted to target the responsible user, for example.

So this logic is not ok

Reply:
Indeed, either you send it to default recipients, or you send it to some particular recipient. We don't have an option "send to a specific recipient or fallback to the default".

If you're already bothering to add placeholders manually in the first place, you can add "or" in there and make your own fallback that will be more accurate than the generic "default recipient" implementation for your use case.

As the ticket seems to simply be "recipients are not getting template emails", it isn't a bug, it's a misconfigured template.